photon-ml/src/main/scala/com/linkedin/photon/ml/util/IOUtils.scala

```diff
@@ -320,12 +329,59 @@ protected[ml] object IOUtils {
       forceOverwrite = true)
   }
 
-  private case class BasicSummaryItems(
-      max: Double,
-      min: Double,
-      mean: Double,
-      normL1: Double,
-      normL2: Double,
-      numNonzeros: Double,
-      variance: Double)
+  /**
+   * Write to a stream while handling exceptions, and closing the stream correctly whether writing to it
+   * succeeded or not.
+   *
+   * NOTE: remember that a Try instance can be understood as a collection, that can have zero
+   * or one element. This code uses a "monadic flow" started by the Try. Try can be a Success or a Failure.
+   * Success.map(lambda) applies lambda to the value wrapped in the Success instance, and returns the result,
+   * which can itself be either Success or Failure, wrapping an instance of the type returned by the lambda.
+   * Failure.map(lambda) ignores lambda, and returns itself, but changing the contained type to be the type
```

Comment:
I don't know if we should document this here or point people to it, but it would be worth noting that a Failure[] always contains an exception cast as whatever type T of Try[T] is.